<?xml version="1.0" ?>
<robot xmlns:xacro="http://ros.org/wiki/xacro">

  <xacro:macro name="simple_arm" params="parent_link *origin">
  <xacro:include filename="$(find irobot_create_description)/urdf/common_properties.urdf.xacro"/>

  <xacro:property name="mass" value="0.00001" />
  <xacro:property name="link_len" value="0.25" />
  <xacro:property name="wrist_len" value="0.05" />
  <xacro:property name="effort_max" value="1000" />
  <xacro:property name="velocity_max" value="10" />
  <xacro:property name="inertial_elem" value="0.001" />


  <joint name="robot_to_base" type="fixed">
    <xacro:insert_block name="origin"/>
    <parent link="${parent_link}"/>
    <child link="arm_base_link"/>
  </joint>

  <link name='arm_base_link'>
    <inertial>
      <mass value="${mass}"/>
      <inertia ixx="${inertial_elem}" ixy="0.0" ixz="0.0" iyy="${inertial_elem}" iyz="0.0" izz="${inertial_elem}"/>
    </inertial>
      <visual name='arm_base_link_visual'>
        <origin xyz="0 0 0"/>
        <geometry>
          <cylinder radius="0.05" length="0.001"/>
        </geometry>
      <material name="grey">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
      </visual>
    </link>

  <joint name="arm_base_joint" type="continuous">
    <axis xyz="0 0 1"/>
    <origin xyz="0.0 0 0.002" rpy="0 0 0"/>
    <parent link="arm_base_link"/>
    <child link="arm_base2_link"/>
    <limit effort="${effort_max}" velocity="${velocity_max}"/>
  </joint>

  <link name='arm_base2_link'>
    <inertial>
      <mass value="${mass}"/>
      <inertia ixx="${inertial_elem}" ixy="0.0" ixz="0.0" iyy="${inertial_elem}" iyz="0.0" izz="${inertial_elem}"/>
    </inertial>
      <visual name='arm_base2_link'>
        <origin xyz="0 0 0"/>
        <geometry>
          <cylinder radius="0.05" length="0.01"/>
        </geometry>
      <material name="grey">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
      </visual>
  </link>

  <joint name="arm_shoulder_joint" type="continuous">
    <axis xyz="0 1 0"/>
    <origin xyz="0.0 0 0.01" rpy="0 0 0"/>
    <parent link="arm_base2_link"/>
    <child link="arm_shoulder_link"/>
    <limit effort="${effort_max}" velocity="${velocity_max}"/>
  </joint>
  
<link name='arm_shoulder_link'>
    <inertial>
      <origin xyz="0 0 ${link_len/2}"/>
      <mass value="${mass}"/>
      <inertia ixx="${inertial_elem}" ixy="0.0" ixz="0.0" iyy="${inertial_elem}" iyz="0.0" izz="${inertial_elem}"/>
    </inertial>
      <visual name='lower_link_visual'>
        <origin xyz="0 0 ${link_len/2}"/>
        <geometry>
          <cylinder radius="0.01" length="${link_len}"/>
        </geometry>
      <material name="grey">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
      </visual>
  </link>

  <joint name="arm_elbow_joint" type="continuous">
    <axis xyz="0 1 0"/>
    <origin xyz="0.0 0 ${link_len}" rpy="0 0 0"/>
    <parent link="arm_shoulder_link"/>
    <child link="arm_elbow_link"/>
    <limit effort="${effort_max}" velocity="${velocity_max}"/>
  </joint>

  <link name='arm_elbow_link'>
    <inertial>
      <origin xyz="0 0 ${link_len/2}"/>
      <mass value="${mass}"/>
      <inertia ixx="${inertial_elem}" ixy="0.0" ixz="0.0" iyy="${inertial_elem}" iyz="0.0" izz="${inertial_elem}"/>
    </inertial>
      <visual name='lower_link_visual'>
        <origin xyz="0 0 ${link_len/2}"/>
        <geometry>
          <cylinder radius="0.01" length="${link_len}"/>
        </geometry>
      <material name="grey">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
      </visual>
  </link>

  <joint name="arm_wrist_joint" type="continuous">
    <axis xyz="0 1 0"/>
    <origin xyz="0.0 0 ${link_len}" rpy="0 0 0"/>
    <parent link="arm_elbow_link"/>
    <child link="arm_wrist_link"/>
    <limit effort="${effort_max}" velocity="${velocity_max}"/>
  </joint>

  <link name='arm_wrist_link'>
    <inertial>
      <origin xyz="0 0 ${wrist_len/2}"/>
      <mass value="${mass}"/>
      <inertia ixx="${inertial_elem}" ixy="0.0" ixz="0.0" iyy="${inertial_elem}" iyz="0.0" izz="${inertial_elem}"/>
    </inertial>
      <visual name='lower_link_visual'>
        <origin xyz="0 0 ${wrist_len/2}"/>
        <geometry>
          <cylinder radius="0.01" length="${wrist_len}"/>
        </geometry>
      <material name="grey">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
      </visual>
  </link>


  <frame name="camera_mount_point">
      <pose relative_to='arm_wrist_link'>0.0 0 0 0 0 0</pose>
  </frame>

  <ros2_control name="arm_controller_hw" type="system">
    <hardware>
      <plugin>ign_ros2_control/IgnitionSystem</plugin>
    </hardware>
    <joint name="arm_base_joint">
      <command_interface name="position"/>
      <state_interface name="position">
        <param name="initial_value">0.0</param>
      </state_interface>
      <state_interface name="velocity"/>
    </joint>
    <joint name="arm_shoulder_joint">
      <command_interface name="position"/>
      <state_interface name="position">
        <param name="initial_value">-0.45</param>
      </state_interface>
      <state_interface name="velocity"/>
    </joint>
    <joint name="arm_elbow_joint">
      <command_interface name="position"/>
      <state_interface name="position">
        <param name="initial_value">2.8</param>
      </state_interface>
      <state_interface name="velocity"/>
    </joint>
    <joint name="arm_wrist_joint">
      <command_interface name="position"/>
      <state_interface name="position">
        <param name="initial_value">-0.8</param>
      </state_interface>
      <state_interface name="velocity"/>
    </joint>
  </ros2_control>

  </xacro:macro>
</robot>
